flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -299,6 +313,15 @@ private NotificationResult fireBufferAvailableNotification(BufferListener listen
 		return notificationResult;
 	}
 
+	/**
+	 * @return true if there is no available buffers in queue and the global quota is also exhausted.
+	 */
+	private boolean isUnavailable() {
+		assert Thread.holdsLock(availableMemorySegments);
+
+		return availableMemorySegments.isEmpty() && numberOfRequestedMemorySegments == currentPoolSize;
```

Comment:
Is it possible that despite `availableMemorySegments.isEmpty() == true` && 
`numberOfRequestedMemorySegments < currentPoolSize;`, we actually should be unavailable? What if there are no available buffers in the global pool?

Reply:
I think it might not be possible to do that directly based on current mechanism. The buffer quota of `LocalBufferPool` should be firstly requested from global pool and then recycled into `availableMemorySegments` after used. In other words, the `availableMemorySegments.isEmpty()` might be always true in the first round but we could request the quota successful from global pool. If we do not try to request from global pool, we do not know the result.

Maybe a feasible way to avoid waiting on requesting buffers from global pool is that the `availableMemorySegments` is filled immediately while constructing the `LocalBufferPool` in eager way, not current lazy way based on emitting data. 

And then we could only rely on `availableMemorySegments.isEmpty()` judging the available state. But it might still exist the case of non available buffers in global pool while constructing the `LocalBufferPool`, so the `LocalBufferPool` should register itself as listeners and global pool would notify the listeners when others return the extra buffers. To do so we could also get ride of the `availableMemorySegments.wait(2000)` in `LocalBufferPool`.

Reply:
> Maybe a feasible way to avoid waiting on requesting buffers from global pool is that the availableMemorySegments is filled immediately while constructing the LocalBufferPool in eager way, not current lazy way based on emitting data.

I think that we can not do, as I think this the mechanism that allows for Flink jobs to run with less then desired amount of buffers (for example only 1 buffer per channel instead of 2).

I was thinking more towards the direction of changing the interaction between `LocalBufferPool` and `NetworkBufferPool`, so that if `LocalBufferPool` is requesting more buffers from the `NetworkBufferPool` and it doesn't get any, it could register some callback/retrieve some `CompletableFuture` that would be completed if some buffers became available (currently `NetworkBufferPool` returns just null).

However there is one more thing. Maybe in the MVP we can live without it. If code will block more often than it could, but the backpressure mechanism based on `isAvailable()` will be still working correctly AND we will not deadlock, that's I think acceptable for now. This current version blocks more often and doesn't deadlock, but I think that currently `isAvailable()` might be completed, while the caller is still blocked on `availableMemorySegments.wait(2000);`.

Reply:
> I think that we can not do, as I think this the mechanism that allows for Flink jobs to run with less then desired amount of buffers (for example only 1 buffer per channel instead of 2)

My above proposal might not effect this behavior. E.g. Assuming we request 100 buffers together from global pool in eager way while constructing the `LocalBufferPool`, and 50 buffers are used in practice and 50 buffers are left in available queue in local pool. If the core size is decreased into 50 because of other new `LocalBufferPool` is constructed, then the 50 available buffers in first local pool could be returned to global pool immediately. If all the 100 buffers in the first local pool are used actually, then even though we request the buffer in lazy way as now, the case is still the same that we have to wait 50 buffers to be recycled into global pool. The benefits of eager way is that we could only rely on the condition of `availableMemorySegments.isEmpty()` to judge the available state at the beginning.

> I was thinking more towards the direction of changing the interaction between LocalBufferPool and NetworkBufferPool, so that if LocalBufferPool is requesting more buffers from the NetworkBufferPool and it doesn't get any, it could register some callback/retrieve some CompletableFuture that would be completed if some buffers became available (currently NetworkBufferPool returns just null).

I approve this way and it has the same effect as the listener way I mentioned before.  And this mechanism would be needed for two purposes:

- We want to judge the available state only by `availableMemorySegments.isEmpty()`.

- We want to get ride of `availableMemorySegments.wait(2000)` in local pool.

> However there is one more thing. Maybe in the MVP we can live without it. If code will block more often than it could, but the backpressure mechanism based on isAvailable() will be still working correctly AND we will not deadlock, that's I think acceptable for now. This current version blocks more often and doesn't deadlock, but I think that currently isAvailable() might be completed, while the caller is still blocked on availableMemorySegments.wait(2000);, which would introduce some issues with the backpressure monitor, right?

ATM the backpressure mechanism still works because we have not changed the processor logic to rely on the output status before processing input. But after we refactor the `StreamTask/StreamInputProcessor` future, that means `StreamTask` would not call `processInput` if the output is unavailable, then we have to change the backpressure monitor. In theory it has to monitor both `RecordWriter#isAvailable` and also `availableMemorySegments.wait(2000)` to cover all the cases.

Reply:
> My above proposal might not effect this behavior

ops, thanks for pointing this and rephrasing your idea.

> I approve this way and it has the same effect as the listener way I mentioned before.

Yes, now I think I get your idea. So in short, we also need to either:

1. add `Future<?> NetworkBufferPool#isAvailable()` method
2. add listeners/callbacks to the `NetworkBufferPool`

I would be in favour of the first option, for the sake of consistency and easier integration with rest of the stack, but if you have opposite preference we can try it out.

> ATM the backpressure mechanism still works 

Yes. I would still consider this `NetworkBufferPool` notifications/future as a required part of this jira ticket. I would slightly prefer to do it as two commits in the same PR, but we can split it as well.

Reply:
> I would be in favour of the first option, for the sake of consistency and easier integration with rest of the stack, but if you have opposite preference we can try it out.

Actually I had no preference before, and the second way was taken in `RemoteInputChannel` requesting floating buffers from `LocalInputChannel`. I am willing to take the first way now because it is widely used recently and seems more unified with other stacks.

> Yes. I would still consider this NetworkBufferPool notifications/future as a required part of this jira ticket. I would slightly prefer to do it as two commits in the same PR, but we can split it as well.

Agree, it can be regarded as the scope of this PR because it also solves another issue of non-blocking output. :)